### PR TITLE
Don't merge constraints, just drop constraints.

### LINKED ISSUE #17599: Bug in AffineConstraints::make_consistent_in_parallel().
Investigating #17595 with @marcfehling has uncovered a problem with merging constraints from different processes in `AffineConstraints::make_consistent_in_parallel()`. In short, here is the situation:

* `DoFTools::make_hanging_node_constraints()` creates constraints for hanging nodes on different processes that *appear* different, because these constraints are created on faces between locally owned and ghost, or ghost and ghost, cells. On some processes, that means that we consider faces that we don't consider on others, and so the constraints we get are different at this stage.
* Specifically, with the short code I attached to #17595 (based on @marcfehling 's testcase), process 0 creates the following constraints among many others:
```
    63 3:  0.323607
    63 7:  -0.123607
    63 19:  0.8
    64 3:  -0.123607
    64 7:  0.323607
    64 19:  0.8
```
* On the other hand, on process 1 we have:
```
   64 3:  -0.447214
    64 7:  0.447214
    64 63:  1
```
* The constraints for 64 *appear* different, but they are not. If you resolve the constraint 64->63 with the constraint we have on process 0, you end up with the exact same constraint as shown for process 0. The issue is that when we merge constraints, we just *add* entries and that leads to substantial confusion.

### LINKED ISSUE #17620: Test lac/constraints_make_consistent_in_parallel_03 is broken.
In my quest to fix #17599, I've realized that the test `lac/constraints_make_consistent_in_parallel_03` is broken. If you look at https://github.com/dealii/dealii/blob/master/tests/lac/constraints_make_consistent_in_parallel_03.cc#L41-L65, you'll realize that the test describes the following situation:
```
proc 0:
  owns 0
  no constraints

proc 1:
  owns 1, 2
  X0 -> 0.5*X1 + 0.5*X2
  X2 -> 1.5*X3

proc 2:
  owns 3, 4
  X0 -> 0.5*X1 + 0.5*X3
```
Resolving the two constraints on proc 1 results in `X= 0.5*X1 + 0.75*X3`, which is not the same as on proc 2. This can't work. I'm going to fix this as part of #17599.

### PR DESCRIPTION
**Pull request by bangerth**
This patch fixes #17599 and fixes #17620 by simply dropping constraints that do not match between processes. As explained in #17599, this is the right approach because non-matching constraints will eventually match up after resolving sufficiently many chains of constraints. 

This PR is currently still WIP because (i) it currently contains half a dozen clean-up commits that I've posted separately, (ii) I still want to add a check that what we ultimately end up with is indeed consistent. I thought it's worth posting anyway so @marcfehling can make progress on other matters.

### CODE REVIEW

**Comment by bangerth** [comment]
There are a number of follow-ups I would like to do, but this is now ready for review and merge. I've added several tests, either specifically created or taken from #17599. I'm pretty sure this is right.

File: include/deal.II/lac/affine_constraints.templates.h [hunk 1]
```diff
@@ -320,149 +320,68 @@ namespace internal
...
+          locally_relevant_constraints.end());
+
+        // For those constraints that survive, sort the right hand side arrays:
+        for (ConstraintLine &entry : locally_relevant_constraints)
+          std::sort(entry.entries.begin(),
+                    entry.entries.end(),
+                    [](const auto &l1, const auto &l2) {
+                      return l1.first < l2.first;
+                    });
       };
```
Comment (marcfehling) [inline include/deal.II/lac/affine_constraints.templates.h:362]: This is the original helper function of release 9.5 before #14905 was added.
https://github.com/dealii/dealii/blob/b7e4db145833d6f74502731fe630296b2a1d910a/include/deal.II/lac/affine_constraints.templates.h#L221-L229

Comparing the algorithm (and ignoring your clean-ups), sorting the right hand side arrays in the helper function was the only change you made.

So we are basically reverting #14905. Do you agree @bergbauer @kronbichler?

**Comment by marcfehling** [comment]
The failing test `trilinos/solver_control_06.mpirun=1.debug` seems unrelated to this PR.

File: include/deal.II/lac/affine_constraints.templates.h (hunk 1 above)
Comment (bergbauer) [reply include/deal.II/lac/affine_constraints.templates.h:362]: Our use case does not work with this patch but it is also broken on master. It used to work with the originally proposed patch together with the old master branch for which it was written. I don't think one can accomplish setting up constraints consistently in parallel for aggregated finite element methods without merging (at least for continuous finite elements). With the test from https://github.com/dealii/dealii/pull/14905 @kronbichler and I tried to imitate what was happening but it seems we have failed at that.

What do you think about the following plan: I open a new issue for setting up constraints consistently for aggregated finite elements where I try to explain the situation and post a test case. In the meantime, we can merge this if it fixes your current problems.

**Review by marcfehling — APPROVED** [review]

File: include/deal.II/lac/affine_constraints.templates.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/lac/affine_constraints.templates.h:362]: @bergbauer  Yes, please do that. Let's first precisely define the semantics you want, and then we can think about how to implement this. I often find it useful to first write the documentation/specification, and then implement what we say there.

**Review by marcfehling — APPROVED** [review]
Let's get this merged!

File: include/deal.II/lac/affine_constraints.h [hunk 2]
```diff
@@ -1994,17 +1994,44 @@ class AffineConstraints : public Subscriptor
...
+   * as second argument to this function what you get from
+   * DoFTools::extract_locally_relevant_dofs(). It is worth noting that the
+   * same happens if you use a continuous Galerkin method but if your
+   * bilinear form contains terms computed on faces -- for example to
+   * penalize certain terms when solving the biharmonic equation (see,
+   * for example, step-47).
+   *
+   * At the end of this function, the current object
+   * only stores constraints for degrees of freedom whose indices are
+   * listed in the second argument. In other words, while
```
Comment (marcfehling) [inline include/deal.II/lac/affine_constraints.h:2028]:
```suggestion
   * listed in @p constraints_to_make_consistent. In other words, while
```
Let's be more precise here. We can address this in a follow-up PR.

Comment (bangerth) [reply include/deal.II/lac/affine_constraints.h:2028]: Yes, I'll add that to my queue for a follow-up. I've got other stuff I want to address anyway. But let's not have to re-start the testers again.

**Pull request merged**